src/connection.ts

```diff
@@ -178,11 +185,28 @@ export class Connection extends JSForceConnection {
    * Retrieves the highest api version that is supported by the target server instance.
    */
   public async retrieveMaxApiVersion(): Promise<string> {
-    type Versioned = { version: string };
-    const versions = (await this.request(`${this.instanceUrl}/services/data`)) as Versioned[];
-    this.logger.debug(`response for org versions: ${versions}`);
-    const max = ensure(maxBy(versions, (version: Versioned) => version.version));
-    return max.version;
+    // errors go to stdout from within jsforce (maybe even further down the stack) if org's url doesn't exist.  (not a 404 from the domain but a DNS ENOTFOUND)
+    // We need to verify that the org can be connected to BEFORE asking about it's api version.  This will throw on DNS errors
+    if (!this.options.connectionOptions?.instanceUrl) {
+      throw new Error('Connection has no instanceUrl');
+    }
+    const resolver = await MyDomainResolver.create({
+      url: new URL(this.options.connectionOptions.instanceUrl),
+      timeout: Duration.minutes(0),
+    });
+    const ipAddress = await resolver.resolve();
+
+    if (ipAddress) {
+      type Versioned = { version: string };
+      const versions = (await this.request(
+        `${this.options.connectionOptions.instanceUrl}/services/data`
+      )) as Versioned[];
+      this.logger.debug(`response for org versions: ${versions}`);
+      const max = ensure(maxBy(versions, (version: Versioned) => version.version));
+      return max.version;
+    }
+
+    throw new Error();
```

Comment:
handled by the caller.  I could find no other references, but because it's public, there should be another bit of text in the Error saying what happened?